data/transactions/application.go

```diff
@@ -0,0 +1,769 @@
+	Accounts        []basics.Address  `codec:"apat,allocbound=256"`
+	ForeignApps     []basics.AppIndex `codec:"apfa,allocbound=256"`
+
+	LocalStateSchema  basics.StateSchema `codec:"apls"`
+	GlobalStateSchema basics.StateSchema `codec:"apgs"`
+	ApprovalProgram   []byte             `codec:"apap"`
+	ClearStateProgram []byte             `codec:"apsu"`
+
+	// If you add any fields here, remember you MUST modify the Empty
+	// method below!
+}
+
+// Empty indicates whether or not all the fields in the
+// ApplicationCallTxnFields are zeroed out
+func (ac *ApplicationCallTxnFields) Empty() bool {
+	if ac.ApplicationID != 0 {
+		return false
+	}
+	if ac.OnCompletion != 0 {
+		return false
+	}
+	if ac.ApplicationArgs != nil {
+		return false
+	}
+	if ac.Accounts != nil {
+		return false
+	}
+	if ac.ForeignApps != nil {
+		return false
+	}
+	if ac.LocalStateSchema != (basics.StateSchema{}) {
+		return false
+	}
+	if ac.GlobalStateSchema != (basics.StateSchema{}) {
+		return false
+	}
+	if ac.ApprovalProgram != nil {
+		return false
+	}
+	if ac.ClearStateProgram != nil {
+		return false
+	}
+	return true
+}
+
+// Allocate the map of LocalStates if it is nil, and return a copy. We do *not*
+// call clone on each AppLocalState -- callers must do that for any values
+// where they intend to modify a contained reference type e.g. KeyValue.
+func cloneAppLocalStates(m map[basics.AppIndex]basics.AppLocalState) map[basics.AppIndex]basics.AppLocalState {
+	res := make(map[basics.AppIndex]basics.AppLocalState, len(m))
+	for k, v := range m {
+		res[k] = v
+	}
+	return res
+}
+
+// Allocate the map of AppParams if it is nil, and return a copy. We do *not*
+// call clone on each AppParams -- callers must do that for any values where
+// they intend to modify a contained reference type e.g. the GlobalState.
+func cloneAppParams(m map[basics.AppIndex]basics.AppParams) map[basics.AppIndex]basics.AppParams {
+	res := make(map[basics.AppIndex]basics.AppParams, len(m))
+	for k, v := range m {
+		res[k] = v
+	}
+	return res
+}
+
+// getAppParams fetches the creator address and AppParams for the app index,
+// if they exist. It does *not* clone the AppParams, so the returned params
+// must not be modified directly.
+func getAppParams(balances Balances, aidx basics.AppIndex) (params basics.AppParams, creator basics.Address, exists bool, err error) {
+	creator, exists, err = balances.GetAppCreator(aidx)
+	if err != nil {
+		return
+	}
+
+	// App doesn't exist. Not an error, but return straight away
+	if !exists {
+		return
+	}
+
+	record, err := balances.Get(creator, false)
+	if err != nil {
+		return
+	}
+
+	params, ok := record.AppParams[aidx]
+	if !ok {
+		// This should never happen. If app exists then we should have
+		// found the creator successfully.
+		err = fmt.Errorf("app %d not found in account %s", aidx, creator.String())
+		return
+	}
+
+	return
+}
+
+func applyDelta(kv basics.TealKeyValue, stateDelta basics.StateDelta) error {
```

Comment:
The order of `stateDelta` entries applies to the `kv` could result in different results. Is that expected and handled by the caller ?

Reply:
That would be very bad if it is possible -- but since there can be at most one value delta per key, I didn't think it was. Is there a scenario I am missing?

Reply:
Given that this is an assumption the function is making, could you please document it at the function header ?

Reply:
Added a comment to the for loop to explain why order doesn't matter (I think it is impossible to construct a `stateDelta` that would apply non-deterministically)